OAuth implementation

moderakh opened the pull request:
This PR:
* adds the foundation for OAuth against Databricks account on AWS with BYOIDP.
* it copies one internal module that Steve Weis @sweisdb wrote for Databricks CLI (oauth.py). Once ecosystem-dev team (Serge, Pieter) build a python sdk core we will move this code to their repo as a dependency. 
* the PR provides authenticators with visitor pattern format for stamping auth-token which later is intended to be moved to the repo owned by Serge @nfx and and Pieter @pietern 

```
connection = sql.connect(
   server_hostname='***.cloud.databricks.com',
   http_path='sql/protocolv1/o/***',
   auth_type="databricks-oauth"
)

for x in range(1, 100):
  cursor = connection.cursor()
  cursor.execute('SELECT * FROM RANGE(10)')
  result = cursor.fetchall()
  for row in result:
    print(row)
  cursor.close()

connection.close()
```

yunbodeng-db commented on src/databricks/sql/auth/__init__.py:
```diff
@@ -0,0 +1,22 @@
+# Databricks CLI
```
Ditto to others.

moderakh replied:
thanks for pointing that out. fixed.

yunbodeng-db commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,103 @@
+def get_authenticator(cfg: ClientContext):
+    if cfg.auth_type == AuthType.DATABRICKS_OAUTH.value:
+        return OAuthAuthenticator(cfg.hostname, cfg.oauth_client_id, cfg.oauth_scopes)
+    elif cfg.token is not None:
```
My concern with adding other authenticators in here too is it adds test burdens (which I cannot see tests being added in the same PR). I would leave the untested (out of scope) code out of the PRs too so we are sure that submitted code has been verified and has good test coverage.

pietern replied:
There are no tests at all at the moment. @moderakh Are you planning to include a couple in this PR?

If so, I recommend keeping these simpler authenticators around.

moderakh replied:
Added tests covering the other auth providers.
https://github.com/databricks/databricks-sql-python/blob/03ab454bdd4d9a39d306873d916fb3496d4496df/tests/test_auth.py

pietern commented on pyproject.toml:
```diff
@@ -13,6 +13,9 @@ python = "^3.7.1"
 thrift = "^0.13.0"
 pyarrow = "^5.0.0"
 pandas = "^1.3.0"
+click=">=7.0"
```
I recommend against depending on click in a non-CLI library.

moderakh replied:
thanks addressed, replaced with logger

pietern commented on src/databricks/sql/auth/__init__.py:
```diff
@@ -0,0 +1,22 @@
+# Databricks CLI
```
These are copied from https://github.com/databricks/databricks-cli. This is not an issue.

pietern commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,103 @@
+                 auth_type: str = None,
+                 use_cert_as_auth: str = None,
+                 tls_client_cert_file: str = None):
+        super().__init__(oauth_scopes="sql offline_access",
```
This should be a constant.

moderakh replied:
addressed

pietern commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,103 @@
+                 tls_client_cert_file: str = None):
+        super().__init__(oauth_scopes="sql offline_access",
+                         # to be changed once registered on the service side
+                         oauth_client_id="databricks-cli",
```
This should be a constant as well.

moderakh replied:
thanks, addressed.

pietern commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,103 @@
+        self.tls_client_cert_file = tls_client_cert_file
+
+
+class SqlConnectorClientContext(ClientContext):
```
What is the value of the subclass here? It doesn't seem to do anything.

moderakh replied:
addressed.

pietern commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+    # I will add support for this outside of the scope of current PR.
+    def __init__(self, hostname, client_id, scopes):
+        self._hostname = self._normalize_host_name(hostname=hostname)
+        self._scope = scopes
```
Mix of singular and plural. If this is an array of scopes, it should be plural.

moderakh replied:
thanks, addressed.

pietern commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+
+from datetime import datetime, timedelta, tzinfo
+
+import click
```
The only use of click here is `click.echo`. Can be replaced with `print`'s to get rid of the dependency.

nfx replied:
print is bad, `logging` must be used - https://docs.python.org/3/howto/logging.html

pietern commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,48 @@
+
+    def flush(self):
+        # TODO retry behaviour
+        # TODO locking for multiple concurrent refresh token
```
Would this client be used in a multi threaded context at all? I would think with methods like `read`, `write`, and `flush`, there is an implied expectation of the caller being single threaded, and multiple threads using multiple clients.

moderakh replied:
thanks for the comment. correct. updated the comment accordingly.

pietern commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,48 @@
+        # TODO locking for multiple concurrent refresh token
+        req = {}
+        self.__authenticator.add_auth_token(req)
+        self._headers['Authorization'] = req['Authorization']
```
Implied dependency of the function setting the `Authorization` header.

This could be more robust by naming the authenticator function `add_headers` and merging the entire return value into the `self._headers` dict. That would also allow for adding multiple headers out of the box.

moderakh replied:
thanks for the comment. addressed.

pietern reviewed (commented):
I suspect you're targeting Python 3 only (as Python 2 is EOL), and if so, this code would benefit from having a couple more type annotations. I thought about this when seeing the `scope` vs `scopes` name. With an annotation, singular or plural would be obvious.

What are your plans for unit tests?

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+
+
+def fetch_well_known_config(idp_url):
+    known_config_url = "{idp_url}/.well-known/oauth-authorization-server".format(idp_url=idp_url)
```
please use f-strings https://realpython.com/python-f-strings/

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+                       "that OAuth is enabled on this account."
+                       .format(status=response.status_code, idp_url=idp_url))
+    try:
+        return json.loads(response.text)
```
use `response.json()`, it's a standard method, re-implementing it here only makes it more complicated

moderakh replied:
addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+    return oauthlib.oauth2.WebApplicationClient(client_id)
+
+
+def get_redirect_url(port=REDIRECT_PORT):
```
what about random port? what if other process is listening on 8020 on client machine?

moderakh replied:
addressed. iterates through the first available port from [8020, 8025)

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+import requests
+from requests.exceptions import RequestException
+
+from databricks_cli.utils import error_and_quit
```
you're not allowed to use it outside of cli package. it calls for `sys.exit`, which is end of the process. use proper exception re-wrapping.

https://github.com/databricks/databricks-cli/blob/main/databricks_cli/utils.py#L94-L100

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+    try:
+        response = requests.request(method="GET", url=known_config_url)
+    except RequestException:
+        error_and_quit("Unable to fetch OAuth configuration from {idp_url}.\n"
```
replace with a dedicated exception, don't use this function, which is equal to panic.

moderakh replied:
the intention was to agree on the rest of code structure before I make changes to oauth class from cli.

already addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+                       "enabled on this account.".format(idp_url=idp_url))
+
+    if response.status_code != 200:
+        error_and_quit("Received status {status} OAuth configuration from "
```
replace with a dedicated exception, don't use this function, which is equal to panic.

moderakh replied:
the intention was to agree on the rest of code structure before I make changes to oauth class from cli.

already addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+
+
+def get_challenge(verifier_string=token_urlsafe(32)):
+    digest = hashlib.sha256(verifier_string.encode('UTF-8')).digest()
+    challenge_string = base64.urlsafe_b64encode(digest).decode("UTF-8").replace('=', '')
```
nit: inconsistent use of single and double quotes in the same method.

moderakh replied:
addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+UTC = UTCTimeZone()
+
+
+def get_client(client_id):
```
this function is used only once and i've mistakenly assumed it as "databricks client". please remove the function and instantiate webapp client directly, for understandability and readability

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+    idp_url = get_idp_url(hostname)
+    oauth_config = fetch_well_known_config(idp_url)
+    token_request_url = oauth_config['token_endpoint']
+    client = get_client(client_id=client_id)
```
```suggestion
    client = oauthlib.oauth2.WebApplicationClient(client_id)
```

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+        # If it has been tampered with, it will be rejected on the server side.
+        # This avoids having to fetch the public key from the issuer and perform
+        # an unnecessary signature verification.
+        decoded = jwt.decode(access_token, options={"verify_signature": False})
```
```suggestion
        decoded = json.loads(base64.standard_b64decode(access_token.split(".")[1]))
```

please rewrite it with something like that - don't depend on `PyJWT`, we don't need it here. jwt tokens are just dot-separated base-64-encoded jsons - https://jwt.io/

moderakh replied:
Thanks @nfx this was very helpful.

nfx commented on pyproject.toml:
```diff
@@ -13,6 +13,9 @@ python = "^3.7.1"
 pyarrow = "^5.0.0"
 pandas = "^1.3.0"
+click=">=7.0"
+pyjwt=">=1.7.0"
```
please remove this dependency.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+    oauth_config = fetch_well_known_config(idp_url)
+    # We are going to override oauth_config["authorization_endpoint"] use the
+    # /oidc redirector on the hostname, which may inject additional parameters.
+    auth_url = "{}oidc/v1/authorize".format(hostname)
```
use f-strings

moderakh replied:
addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+            challenge,
+            REDIRECT_PORT)
+    except OAuth2Error as err:
+        error_and_quit("OAuth Authorization Error: {error}".format(error=err.description))
```
rewrap as proper exceptions

moderakh replied:
thanks already addressed.

nfx commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+            state,
+            challenge,
+            REDIRECT_PORT)
+    except OAuth2Error as err:
```
```suggestion
    except OAuth2Error as e:
```

is more idiomatic way of catching exceptions in python

moderakh replied:
thanks addressed.

nfx commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,48 @@
+        # TODO retry behaviour
+        # TODO locking for multiple concurrent refresh token
+        req = {}
+        self.__authenticator.add_auth_token(req)
+        self._headers['Authorization'] = req['Authorization']
```
please use visitor pattern or anything else, that will set more than one header. most of the authorizers do set more than one headers, sometimes three.

moderakh replied:
thanks for pointing this out. addressed.

nfx commented on src/databricks/sql/client.py:
```diff
@@ -90,25 +85,7 @@ def __init__(
-            raise ValueError(
-                "No valid authentication settings. Please provide an access token."
-            )
+        authenticator = get_python_sql_connector_authenticator(server_hostname, **kwargs)
```
call it `auth`. or visitor. not authenticator, because it's not an established name in the unified client auth.

moderakh replied:
addressed, renamed to auth_provider

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+import base64
+
+
+class Authenticator:
```
```suggestion
# Private API: this is an evolving interface and it will change in the future.
# Please must not depend on it in your applications.
class CredentialsProvider:
```

(or `Authorizer` - authenticator is something different)

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+        pass
+
+
+class AccessTokenAuthenticator(Authenticator):
```
```suggestion
# Private API: this is an evolving interface and it will change in the future.
# Please must not depend on it in your applications.
class AccessTokenAuthenticator(Authenticator):
```

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+        request_headers['Authorization'] = self.__authorization_header_value
+
+
+class UserPassAuthenticator(Authenticator):
```
```suggestion
# Private API: this is an evolving interface and it will change in the future.
# Please must not depend on it in your applications.
class BasicAuthenticator(Authenticator):
```

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+        request_headers['Authorization'] = self.__authorization_header_value
+
+
+class OAuthAuthenticator(Authenticator):
```
```suggestion
# Private API: this is an evolving interface and it will change in the future.
# Please must not depend on it in your applications.
class DatabricksOAuthAuthenticator(Authenticator):
```

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+    # TODO: moderakh the refresh_token is only kept in memory. not saved on disk
+    # hence if application restarts the user may need to re-authenticate
+    # I will add support for this outside of the scope of current PR.
+    def __init__(self, hostname, client_id, scopes):
```
hostname should not be a property of credentials provider and it's to be taken from configuration.

moderakh replied:
@nfx I am not sure if I understood what you meant in this comment.

`DatabricksOAuthProvider` hides the oauth implementation, so it needs to be aware of hostname or redirect url to initiate oauth protocol. 

in the python sql connector, the end point information is passed as following
```python 
connection = sql.connect(
      server_hostname='moderakh-test.cloud.databricks.com',
      http_path='sql/protocolv1/o/***',
      auth_type="databricks-oauth"
)
```

the implementation passed `server_hostname` down to `DatabricksOAuthProvider` to be used by oauth implementation, there is no independent configuration support.

nfx commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,78 @@
+        request_headers['Authorization'] = "Bearer {}".format(self._access_token)
+
+    @staticmethod
+    def _normalize_host_name(hostname):
```
this belongs to the layer that instantiates this class, not this concrete class

moderakh replied:
done.

nfx reviewed (changes requested):
code smell: all functions of `src/databricks/sql/auth/oauth.py` are shuffled to a top-level namespace and it's difficult to understand which methods of  belong to Authorizer and which are related to web browser flow. please refactor this file to be a bit more object-oriented.

please remove `click` and `pyjwt` dependencies - you don't need them. perhaps even `databricks_cli` as dependency.

moderakh commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+        state=state,
+        code_challenge=challenge,
+        code_challenge_method="S256")
+    click.echo("Opening {uri}".format(uri=auth_req_uri))
```
addressed.

moderakh commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+
+def send_auth_code_token_request(client, token_request_url, redirect_url, code, verifier):
+    token_request_body = client.prepare_request_body(code=code, redirect_uri=redirect_url)
+    data = "{body}&code_verifier={verifier}".format(body=token_request_body, verifier=verifier)
```
addressed.

moderakh commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+                       .format(uri=redirect_url))
+    # This is a kludge because the parsing library expects https callbacks
+    # We should probably set it up using https
+    full_redirect_url = "https://localhost:{port}/{path}".format(
```
addressed.

moderakh commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,286 @@
+
+
+def get_authorization_code(client, auth_url, redirect_url, scope, state, challenge, port):
+    #pylint: disable=unused-variable
```
addressed.

moderakh commented:
> @pietern  I suspect you're targeting Python 3 only (as Python 2 is EOL), and if so, this code would benefit from having a couple more type annotations. I thought about this when seeing the scope vs scopes name. With an annotation, singular or plural would be obvious.
> What are your plans for unit tests?

thanks for the comment.
* We only support Python 3.
* I already added type annotation for scopes (thanks for pointing that out, will try to add for others as well). 
* I already added some unit test for other auth providers (access-token, user/pass here: https://github.com/databricks/databricks-sql-python/blob/410ef1642339aba1f681501b9911f57a4fd71ed6/tests/test_auth.py
* adding test for oauth will be a bit more complex, probably will add some mocking test, but that will be the last piece as I need everything else to be correctly structured to not do re-do refactor tests (I may do that one after this PR merges)

moderakh commented:
> @nfx code smell: all functions of src/databricks/sql/auth/oauth.py are shuffled to a top-level namespace and it's difficult to understand which methods of belong to Authorizer and which are related to web browser flow. please refactor this file to be a bit more object-oriented.
> 
> please remove click and pyjwt dependencies - you don't need them. perhaps even databricks_cli as dependency.

@nfx Thanks for the comment. I removed pyjwt and click dependencies. I will look at refactoring the code in oauth.py that we borrowed from databricks-cli. except this one and one or two other comments, all of your other comments are addressed.

sweisdb commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,282 @@
+logger = logging.getLogger(__name__)
+
+
+# This could use 'import secrets' in Python 3
```
Do we need to support Python 2? If not, we can drop some of these workarounds for Python 2.

moderakh replied:
Thanks for the comment. only python 3. I changed the code to use python 3 APIs. addressed here and in one other place.

sweisdb commented on src/databricks/sql/auth/oauth.py:
```diff
@@ -0,0 +1,282 @@
+
+class SingleRequestHandler(BaseHTTPRequestHandler):
+    RESPONSE_BODY = """<html>
+<head>
```
Maybe put this into an external file where it will be easier to view and edit? (Yes, I know this is what we did for databricks-cli, but that was kind of treated as experimental.)

moderakh replied:
thanks for the comment addressed. also I got rid of the global variable.

sweisdb reviewed (approved):
LGTM overall.

pietern commented on src/databricks/sql/__init__.py:
```diff
@@ -44,7 +44,6 @@ def TimestampFromTicks(ticks):
     from .client import Connection
-
-    return Connection(server_hostname, http_path, access_token, **kwargs)
+    return Connection(server_hostname, http_path, experimental_oauth_persistence, **kwargs)
```
This file can be reverted; no need to call out `experimental_oauth_persistence` here as it is carried forward by `**kwargs` just like all the other optional keyword arguments.

moderakh replied:
thanks for the comment. addressed.

pietern commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,96 @@
+
+
+OAUTH_SCOPES = ["sql", "offline_access"]
+# TODO: moderakh to be changed once registered on the service side
```
Does this need to be addressed prior to merging?

moderakh replied:
thanks addressed. oauth client-id and redirect urls are registered in the service side. addressed.

pietern commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,46 @@
+        super().setCustomHeaders(headers)
+
+    def flush(self):
+        # TODO retry behaviour
```
If this is a real TODO, please refer to issue # here. Otherwise it should be removed.

susodapop replied:
I think retry behaviour is handled in `thrift_backend`, no?

moderakh replied:
not a real TODO. removed thanks.

pietern commented on src/databricks/sql/client.py:
```diff
@@ -26,7 +22,7 @@ def __init__(
         server_hostname: str,
         http_path: str,
-        access_token: str,
+        oauth_persistence: OAuthPersistence = None,
```
Can you add back `access_token: str = None` here? Consistent with the documentation below.

If you don't want to mention authorization related arguments here but rather keep them all in `**kwargs`, both `access_token` and `oauth_persistence` shouldn't be mentioned. Having one but not the other is inconsistent.

moderakh replied:
thanks for the comment moved both to `**kwargs`

pietern commented on src/databricks/sql/client.py:
```diff
@@ -90,25 +86,7 @@ def __init__(
-            raise ValueError(
-                "No valid authentication settings. Please provide an access token."
-            )
+        auth_provider = get_python_sql_connector_auth_provider(server_hostname, oauth_persistence, **kwargs)
```
Per command above, either pass all arguments explicitly, or rely on `**kwargs` for all.

moderakh replied:
addressed. thanks.

pietern commented on src/databricks/sql/experimental/oauth_persistence.py:
```diff
@@ -0,0 +1,56 @@
+            outfile.write(json_object)
+
+    def read(self) -> OAuthToken:
+        # TODO: validate the
```
Broken comment.

pietern commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,96 @@
+OAUTH_CLIENT_ID = "databricks-sql-python"
+OAUTH_REDIRECT_PORT_RANGE = range(8020, 8025)
+
+def get_python_sql_connector_auth_provider(hostname: str, oauth_persistence: OAuthPersistence = None, **kwargs):
```
Per the other comments,  `oauth_persistence` should be lifted from `**kwargs` as well.

moderakh replied:
thanks, addressed.

pietern reviewed (commented):
Couple of comments around `**kwargs` vs explicitly named keyword arguments, for consistency.

susodapop commented on src/databricks/sql/auth/oauth_http_handler.py:
```diff
@@ -0,0 +1,42 @@
+from http.server import BaseHTTPRequestHandler
```
I expect some of this will be explained in the design doc: but why does `databricks-sql-connector` need to run an HTTP server? Is this for the browser to redirect a callback URL after the user successfully authenticates?

moderakh replied:
correct. Interactive User to Machine OAuth code flow requires the oauth code to be returned on the redirect url where an http server is listening. This class provides the implementation which listens for oauth code on the configured redirect url (which should be http://localhsot:8020 or other ports if this one taken.)

susodapop commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,130 @@
+
+class CredentialsProvider:
+    def add_headers(self, request_headers: Dict[str, str]):
+        pass
```
Ditto previous comment. 
```suggestion
        raise NotImplementedError()
```

moderakh replied:
ditto as above.

This is actually not meant to be an interface but a base class with no op.

what's the use case?
the use case is when the existing `_use_cert_as_auth` and `_tls_client_cert_file` are used, in that scenario we don't add any auth additional header to the request.

susodapop commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,46 @@
+
+import logging
+
+import thrift.transport.THttpClient
```
Nit: It feels strange to use the fully qualified name in two places (here and line 30). You can instead import `thrift` here.
```suggestion
import thrift
```

or 
```suggestion
from thrift.transport.THttpClient import THttpClient
```

susodapop commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,46 @@
+logger = logging.getLogger(__name__)
+
+
+class THttpClient(thrift.transport.THttpClient.THttpClient):
```
Nit: no type hints for this Class?

moderakh replied:
thanks addressed.

moderakh commented on src/databricks/sql/experimental/oauth_persistence.py:
```diff
@@ -0,0 +1,56 @@
+    def __init__(self, access_token, refresh_token):
+        self._access_token = access_token
+        self._refresh_token = refresh_token
+
+    def get_access_token(self) -> str:
+        return self._access_token
```
Thanks @susodapop this is very helpful.

moderakh commented on src/databricks/sql/experimental/oauth_persistence.py:
```diff
@@ -0,0 +1,56 @@
+
+class OAuthPersistence:
+    def persist(self, oauth_token: OAuthToken):
+        pass
```
This is actually not meant to be an interface but a base class with no op.

what's the use case?
the use case is when the existing `_use_cert_as_auth` and `_tls_client_cert_file` are used, in that scenario we don't add any auth additional header to the request.

moderakh commented:
> It's unusual for files in this repository to all include the APACHE licence since this is already covered by our ./LICENSE file.

Thanks @susodapop , I removed the Apache license from the new files to be consistent with the existing files. Addressed.

pietern commented on src/databricks/sql/__init__.py:
```diff
@@ -44,7 +44,7 @@ def TimestampFromTicks(ticks):
 
 
-def connect(server_hostname, http_path, access_token, **kwargs):
+def connect(server_hostname, http_path, **kwargs):
```
If folks have called this function without keyword arguments in the path, e.g.
```python
connect("hostname", "path", "token")
```

This modification would break that.

Might be fine, but worth checking. @susodapop WDYT?

moderakh replied:
thanks @pietern for catching this. will fix it shortly.

moderakh replied:
addressed

pietern reviewed (approved):
Last comment, rest looks good. Accepting to unblock progress.

nfx reviewed (approved):
LGTM 👍

benfleis commented on src/databricks/sql/__init__.py:
```diff
@@ -44,7 +44,6 @@ def TimestampFromTicks(ticks):
 
 
-def connect(server_hostname, http_path, access_token, **kwargs):
+def connect(server_hostname, http_path, access_token=None, **kwargs):
```
Thinking out loud -- this param being called `access_token` may conflate it with OAuth access tokens, and possibly appear to exclude PAT tokens.

If the earlier assertion is correct that it's used as a required positional param, is this the ideal moment to rename to `auth_token` to avoid the PAT conflation and make consistent with the attached docs/logs?

sander-goos commented on src/databricks/sql/auth/authenticators.py:
```diff
@@ -0,0 +1,118 @@
+from databricks.sql.experimental.oauth_persistence import OAuthToken, OAuthPersistence
+
+
+class CredentialsProvider:
```
nit: would `AuthProvider` be more consistent with the other names? The implementations all have this as suffix, and the field in which it is stored is also called: `auth_provider`

moderakh replied:
renamed.

sander-goos commented on src/databricks/sql/auth/thrift_http_client.py:
```diff
@@ -0,0 +1,35 @@
+    def flush(self):
+        headers = dict(self._headers)
+        self.__auth_provider.add_headers(headers)
+        self._headers = headers
```
nit: setting `self._headers` is also done by `setCustomHeaders`

moderakh replied:
`self.__auth_provider.add_headers()` adds the auth headers. `headers` here contains auth token and is not the same as what was set by the `setCustomHeaders()`

sander-goos replied:
I meant that `self.setCustomHeaders(self._headers)` is called at line 34 (after the auth headers have been added), which also sets `self._headers`

sander-goos reviewed (approved):
Great stuff, thanks for implementing this! Added some nits/questions

moderakh commented on src/databricks/sql/client.py:
```diff
@@ -36,15 +33,77 @@ def __init__(
+                This is currently in private preview for Databricks accounts on AWS.
+                This is a beta feature and supports User to Machine OAuth authentication for Databricks on AWS with
+                any IDP configured. This is only for interactive python applications and open a browser window.
+                Note this is beta (private preview)
```
Thanks addressed.

moderakh commented on src/databricks/sql/client.py:
```diff
@@ -36,15 +33,77 @@ def __init__(
+                any IDP configured. This is only for interactive python applications and open a browser window.
+                Note this is beta (private preview)
+
+            experimental_oauth_persistence: configures preferred storage for persisting oauth tokens.
```
We do not have dependency on Databricks CLI. We rely on the end user to implement the token persistence layer. ecosystem dev team is planning to build configuration management in the python core sdk. once that's ready we can also use that as option.

moderakh commented on src/databricks/sql/auth/auth.py:
```diff
@@ -0,0 +1,96 @@
+PYSQL_OAUTH_REDIRECT_PORT_RANGE = list(range(8020, 8025))
+
+
+def normalize_host_name(hostname: str):
```
that one slightly different constructs the oidc url

sander-goos replied:
I see. On a second look, this doesn't take into account the `port` (or `_connection_uri`) parameters. (Also, we can assume `server_hostname` never contains a scheme or trailing slash as otherwise the connection will fail: https://github.com/databricks/databricks-sql-python/blob/main/src/databricks/sql/thrift_backend.py#L105-L107 )

Brings up another question: why is this function actually needed at this point? It seems confusing as the field name is called `hostname` whereas it really is storing a uri. The usages of this field in `DatabricksOAuthProvider` (oauth_persistence, oauth_manager) seem to work fine with just a hostname (although again they would be missing the `port`).
